mava/components/jax/building/system_init.py

```diff
@@ -0,0 +1,247 @@
+from dataclasses import dataclass, field
+from typing import Callable, List, Optional, Union
+
+from mava.components.jax import Component
+from mava.core_jax import SystemBuilder
+from mava.utils import enums
+from mava.utils.sort_utils import sample_new_agent_keys, sort_str_num
+
+
+@dataclass
+class FixedNetworkSystemInitConfig:
+    shared_weights: bool = True
+    network_sampling_setup: enums.NetworkSampler = (
+        enums.NetworkSampler.fixed_agent_networks
+    )
+
+
+class FixedNetworkSystemInit(Component):
+    def __init__(
+        self, config: FixedNetworkSystemInitConfig = FixedNetworkSystemInitConfig()
+    ):
+        """Agents will always be assigned to a fixed network.
+
+        A single network get assigned to all agents of the same type
+        if weights are shared, else a seperate network is assigned to each agent
+
+        Args:
+            config : a dataclass specifying the component parameters
+        """
+        self.config = config
+
+    def on_building_init(self, builder: SystemBuilder) -> None:
+        """Compute and add network sampling information to the builder."""
+
+        if (
+            self.config.network_sampling_setup
+            != enums.NetworkSampler.fixed_agent_networks
+        ):
+            raise ValueError(
+                "Fixed network system init requires fixed_agent_networks sampling"
+            )
+
+        builder.store.agent_net_keys = {
+            agent: f"network_{agent.split('_')[0]}"
+            if self.config.shared_weights
+            else f"network_{agent}"
+            for agent in builder.store.agents
+        }
+        builder.store.network_sampling_setup = [
+            [
+                builder.store.agent_net_keys[key]
+                for key in sort_str_num(builder.store.agent_net_keys.keys())
+            ]
+        ]
+
+        # Check that the environment and agent_net_keys has the same amount of agents
+        sample_length = len(builder.store.network_sampling_setup[0])
+        agent_ids = builder.store.environment_spec.get_agent_ids()
+        assert len(agent_ids) == len(builder.store.agent_net_keys.keys())
+
+        # Check if the samples are of the same length and that they perfectly fit
+        # into the total number of agents
+        assert len(builder.store.agent_net_keys.keys()) % sample_length == 0
+        for i in range(1, len(builder.store.network_sampling_setup)):
+            assert len(builder.store.network_sampling_setup[i]) == sample_length
+
+        # Get all the unique agent network keys
+        all_samples = []
+        for sample in builder.store.network_sampling_setup:
+            all_samples.extend(sample)
+        builder.store.unique_net_keys = list(sort_str_num(list(set(all_samples))))
+
+        # Create mapping from ints to networks
+        builder.store.net_keys_to_ids = {
+            net_key: i for i, net_key in enumerate(builder.store.unique_net_keys)
+        }
+
+    @staticmethod
```

Comment:
I think we should create a baseSystemInit for the 3 classes with these shared methods (as done with the adder components)